Add Schwarz weighting functions

nilsvu opened the pull request:
## Proposed changes

Weighting functions for the additive Schwarz solver, used to combine solutions of overlapping element-centered subdomains.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add Schwarz weighting functions

nilsvu commented:
@wthrowe @moxcodes could you review this one? It uses the hypercube iterators that you reviewed recently.

wthrowe commented on src/ParallelAlgorithms/LinearSolver/Schwarz/Weighting.hpp:
```diff
@@ -0,0 +1,131 @@
+ *
+ * \f{equation}
+ * w(\xi) = \frac{1}{2}\left( \phi\left( \frac{\xi + 1}{\delta} \right) -
+ * \phi(\left( \frac{\xi - 1}{\delta} \right)) \right) \f}
```
Too many parens.

wthrowe commented on src/ParallelAlgorithms/LinearSolver/Schwarz/Weighting.cpp:
```diff
@@ -0,0 +1,214 @@
+  return extruding_weight(logical_coords - sign * 2., width, opposite(side));
+}
+
+size_t dim_in_volume(const size_t dim_in_slice,
```
Put in anonymous namespace.

wthrowe commented on tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_Weighting.cpp:
```diff
@@ -0,0 +1,134 @@
+namespace LinearSolver::Schwarz {
+
+template <size_t Dim>
+void test_weights_conservation(const Element<Dim>& element,
```
Anonymous namespace.

nilsvu commented:
Thanks @wthrowe, addressed your comments and squashed them directly since they were quite small.

nilsvu commented:
@moxcodes would you be able to review this one? If you can't please let me know then I'll ask another reviewer :)

moxcodes commented:
Yep, I'll review within a couple of days.

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Schwarz/Weighting.hpp:
```diff
@@ -0,0 +1,131 @@
+ * w(\xi) = \frac{1}{2}\left( \phi\left( \frac{\xi + 1}{\delta} \right) -
+ * \phi\left( \frac{\xi - 1}{\delta} \right) \right) \f}
+ *
+ * where \f$\phi(\xi)\f$ is a second-order smoothstep, i.e. the quintic
```
refer to the `smoothstep` function you call so it's clear where to find the implementation.

moxcodes commented on src/ParallelAlgorithms/LinearSolver/Schwarz/Weighting.cpp:
```diff
@@ -0,0 +1,216 @@
+        get(edge_weight) *=
+            intruding_weight(logical_coords.get(d_perp),
+                             gsl::at(overlap_widths, d_perp), edge.side());
+        get(*weight) += get(edge_weight) / (Dim - 1);
```
For my understanding, can you explain how the edge weight interacts with the corner weight values -- how do you avoid overcounting?

nilsvu replied:
First, edge-neighbors only exist in 3D, so let's assume that. Now in 3D I iterate over corners and edges separately, first adding intruding weights from corner-neighbors and then adding intruding weights from edge-neighbors. These are distinct elements (a corner-neighbor can't be an edge-neighbor and vice versa), so there's no overcounting.

moxcodes commented on tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_Weighting.cpp:
```diff
@@ -0,0 +1,136 @@
+                            logical_coords, direction, overlap_widths,
+                            num_neighbors, element.external_boundaries()));
+  }
+  CHECK_ITERABLE_APPROX(get(summed_weights), DataVector(num_points, 1.));
```
Is it true that if the weights sum to 1 that all of the corner logic has been correctly applied? If it's possible to mess up the corners and still get 1, then it's probably wise to have some simple 2 or 3 D case worked explicitly; e.g. a 3x3x3 element with a convenient choice for overlap width.

nilsvu replied:
I think so, yes. I test the weights on an element with all possible configurations of neighbors, so the conservation test is pretty thorough. However, looking at the tests again I found one possible failure mode that I added another test for: The exact shape of the weighting function wasn't fixed by the tests beyond requiring it goes from 0 to 1 and is 0.5 at the center, so I now test it explicitly by checking the weighting function shape against Eqn. (41) in https://arxiv.org/abs/1907.01572.

nilsvu committed:
Fix a bug in smoothstep (or Blaze)

The blaze::clamp function returns DataVectors with wrong values
in Release builds when the argument is a un-evaluated Blaze expression.

nilsvu commented:
@moxcodes Thanks for reviewing, I addressed your comments. Please take another look.

moxcodes commented:
Looks good, go ahead and squash.

nilsvu commented:
Thanks for your reviews @wthrowe and @moxcodes! I think this is ready to go whenever you get the chance @wthrowe :)

nilsvu commented:
@wthrowe could you take a look? Seems ready to go
